server/src/main/java/org/apache/iotdb/db/engine/compaction/writer/AbstractCrossCompactionWriter.java

```diff
@@ -0,0 +1,223 @@
+    this.deviceId = deviceId;
+    this.isAlign = isAlign;
+    this.seqFileIndexArray = new int[subTaskNum];
+    checkIsDeviceExistAndGetDeviceEndTime();
+    for (int i = 0; i < targetFileWriters.size(); i++) {
+      chunkGroupHeaderSize = targetFileWriters.get(i).startChunkGroup(deviceId);
+    }
+  }
+
+  @Override
+  public void endChunkGroup() throws IOException {
+    for (int i = 0; i < seqTsFileResources.size(); i++) {
+      TsFileIOWriter targetFileWriter = targetFileWriters.get(i);
+      if (isDeviceExistedInTargetFiles[i]) {
+        targetFileWriter.endChunkGroup();
+      } else {
+        targetFileWriter.truncate(targetFileWriter.getPos() - chunkGroupHeaderSize);
+      }
+      isDeviceExistedInTargetFiles[i] = false;
+    }
+    seqFileIndexArray = null;
+  }
+
+  @Override
+  public void endMeasurement(int subTaskId) throws IOException {
+    flushChunkToFileWriter(
+        targetFileWriters.get(seqFileIndexArray[subTaskId]), chunkWriters[subTaskId]);
+    seqFileIndexArray[subTaskId] = 0;
+  }
+
+  @Override
+  public void write(long timestamp, Object value, int subTaskId) throws IOException {
+    checkTimeAndMayFlushChunkToCurrentFile(timestamp, subTaskId);
+    int fileIndex = seqFileIndexArray[subTaskId];
+    writeDataPoint(timestamp, value, chunkWriters[subTaskId]);
+    chunkPointNumArray[subTaskId]++;
+    checkChunkSizeAndMayOpenANewChunk(
+        targetFileWriters.get(fileIndex), chunkWriters[subTaskId], subTaskId, true);
+    isDeviceExistedInTargetFiles[fileIndex] = true;
+    isEmptyFile[fileIndex] = false;
+  }
+
+  /** Write data in batch, only used for aligned device. */
+  @Override
+  public abstract void write(TimeColumn timestamps, Column[] columns, int subTaskId, int batchSize)
+      throws IOException;
+
+  @Override
+  public void endFile() throws IOException {
+    for (int i = 0; i < isEmptyFile.length; i++) {
+      targetFileWriters.get(i).endFile();
+      // delete empty target file
+      if (isEmptyFile[i]) {
+        targetFileWriters.get(i).getFile().delete();
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    for (TsFileIOWriter targetWriter : targetFileWriters) {
+      if (targetWriter != null && targetWriter.canWrite()) {
+        targetWriter.close();
+      }
+    }
+    targetFileWriters = null;
+    seqTsFileResources = null;
+  }
+
+  @Override
+  public void checkAndMayFlushChunkMetadata() throws IOException {
+    for (int i = 0; i < targetFileWriters.size(); i++) {
+      TsFileIOWriter fileIOWriter = targetFileWriters.get(i);
+      TsFileResource resource = targetResources.get(i);
+      // Before flushing chunk metadatas, we use chunk metadatas in tsfile io writer to update start
+      // time and end time in resource.
+      List<TimeseriesMetadata> timeseriesMetadatasOfCurrentDevice =
+          fileIOWriter.getDeviceTimeseriesMetadataMap().get(deviceId);
```

Comment:
Calling this function is costly.

Reply:
Resolved. I have added a new method named `getChunkMetadatasOfDeviceInMemory()` in `TsFileIOWriter`.